src/main/java/org/apache/maven/shared/transfer/dependencies/collect/internal/Maven30DependencyNodeAdapter.java

```diff
@@ -0,0 +1,164 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.maven.RepositoryUtils;
+import org.apache.maven.artifact.Artifact;
+import org.apache.maven.artifact.repository.ArtifactRepository;
+import org.apache.maven.shared.transfer.dependencies.collect.DependencyCollectorException;
+import org.apache.maven.shared.transfer.graph.DependencyNode;
+import org.apache.maven.shared.transfer.graph.DependencyVisitor;
+import org.sonatype.aether.repository.RemoteRepository;
+
+/**
+ * DependencyCollectorNode wrapper around {@link org.sonatype.aether.graph.DependencyNode}
+ * 
+ * @author Pim Moerenhout
+ *
+ */
+class Maven30DependencyNodeAdapter implements DependencyNode
+{
+
+    private org.sonatype.aether.graph.DependencyNode dependencyNode;
+
+    /**
+     * @param dependencyNode {@link org.sonatype.aether.graph.DependencyNode}
+     */
+    Maven30DependencyNodeAdapter( org.sonatype.aether.graph.DependencyNode dependencyNode )
+    {
+        this.dependencyNode = dependencyNode;
+    }
+
+    @Override
+    public Artifact getArtifact()
+    {
+        return getArtifact( dependencyNode.getDependency().getArtifact() );
+    }
+
+    @Override
+    public List<DependencyNode> getChildren()
+    {
+        List<org.sonatype.aether.graph.DependencyNode> aetherChildren = dependencyNode.getChildren();
+        List<DependencyNode> children = new ArrayList<>( aetherChildren.size() );
+        for ( org.sonatype.aether.graph.DependencyNode aetherChild : aetherChildren )
+        {
+            children.add( new Maven30DependencyNodeAdapter( aetherChild ) );
+        }
+        return children;
+    }
+
+    @Override
+    public List<ArtifactRepository> getRemoteRepositories()
+    {
+        List<RemoteRepository> aetherRepositories = dependencyNode.getRepositories();
+        List<ArtifactRepository> mavenRepositories = new ArrayList<ArtifactRepository>( aetherRepositories.size() );
+
+        for ( RemoteRepository aetherRepository : aetherRepositories )
+        {
+            mavenRepositories.add( new Maven30ArtifactRepositoryAdapter( aetherRepository ) );
+        }
+
+        return mavenRepositories;
+    }
+
+    @Override
+    public Boolean getOptional()
+    {
+        return dependencyNode.getDependency().isOptional();
+    }
+
+    @Override
+    public String getScope()
+    {
+        return dependencyNode.getDependency().getScope();
+    }
+
+    @Override
+    public boolean accept( DependencyVisitor visitor )
+    {
+        if ( visitor.visitEnter( this ) )
+        {
+            List<org.sonatype.aether.graph.DependencyNode> aetherNodes = dependencyNode.getChildren();
+            for ( int i = 0, n = aetherNodes.size(); i < n; i++ )
```

Comment:
Why not for each?

Reply:
Indeed, replaced with for().

Reply:
I'm pretty sure @michael-o was expecting this:

    for ( org.sonatype.aether.graph.DependencyNode child : dependencyNode.getChildren() )
    {
         DependencyNode node = new Maven30DependencyNodeAdapter( child );
         if ( !node.accept( visitor ) )
         {
             break;
         }
    } 

Easier to read and probably more efficient.

Reply:
Implemented the expected!